File: ttrss_update.rb
#!/usr/bin/env ruby
require 'pathname'
require 'fileutils'
require 'optparse'

require_relative 'andyrb/util'
require_relative 'andyrb/git'
require_relative 'andyrb/mood'

class Options
  def self.parse(args)
    options = OpenStruct.new
    options.aggressive = false
    options.replace = false

    optparse = OptionParser.new do |opts|
      opts.on('-a', '--aggressive', 'Activate git gc in aggressive mode if run.') { options.aggressive = true }
      opts.on('-r', '--replace', 'Rename any existing directory if a repository is not found.') { options.replace = true }
    end
    optparse.parse!(args)
    options
  end
end

ARGV.compact! if ARGV.respond_to?(:compact!)
ARGV.flatten! if ARGV.respond_to?(:flatten!)
ARGV.uniq! if ARGV.respond_to?(:uniq!)
Args = Options.parse(ARGV)
TTrss = Git.new('/data/web/feeds', use_sudo: True, sudo_user: 'nginx')
at_exit do
  TTrss.clean_lock
end

def ttrssgcfile
  cachepath = Pathname.new('/var/cache/ttrssgc')
  use_sudo = false
  use_sudo = true unless cachepath.parent.writable?
  case
  when use_sudo
    Util::Program.runprogram(%W(touch #{cachepath}), use_sudo: true)
  else
    FileUtils.touch(cachepath.to_s)
  end
end

ttrssgcfile unless File.exist?('/var/cache/ttrssgc')

def ttrssgc(thendate)
  diff = Util::Datediff.getdiff(thendate)
  TTrss.gc(Args.aggressive) if diff >= 30
end

repopath = Pathname.new('/data/web/feeds/.git')
wdpath = repopath.parent
giturl = 'https://tt-rss.org/gitlab/fox/tt-rss.git'
case
when repopath.directory?
  TTrss.clean_lock
  TTrss.pull
when wdpath.exist? && !repopath.exist?
  case
  when Args.replace
    puts Mood.neutral("No git repository located in the target directory, renaming it to #{wdpath}.old")
    Util::Program.runprogram(%W(mv #{wdpath} #{wdpath}.old), use_sudo: true, sudo_user: 'nginx')
    TTrss.clone(giturl)
  else
    puts Mood.sad('No git repository located in the target directory.')
    raise
  end
when wdpath.file?
  puts Mood.neutral('Target location is a file, renaming and cloning repository.')
  Util::Program.runprogram(%W(mv #{wdpath} #{wdpath}.bad), use_sudo: true, sudo_user: 'nginx')
  TTrss.clone(giturl)
else
  puts Mood.neutral('Target location does not exist, cloning repository')
  TTrss.clone(giturl)
end

ttrssgc(File.stat('/var/cache/ttrssgc').mtime)

File: ttrss_update.py
#!/usr/bin/env python3
# pylint: disable=line-too-long
import argparse
import atexit
import pathlib
import shlex

from andypy.git import Git
# from andypy.util import Mood, Program, Util
from andypy.mood2 import Mood
from andypy.program import Program
from andypy.util.datediff import datediff
from andypy.util.touchfile import touchfile

args = argparse.ArgumentParser()
args.add_argument("--aggressive", "-a", action="store_true", help="Run garbage collection in aggressive mode")
args.add_argument("--replace", "-r", action="store_true", help="Rename any existing directory if a repository is not found.")
args.add_argument("--restart-ttrssd", action="store_true", dest="restart", help="Restart ttrssd")
options = vars(args.parse_args())

ttrss = Git("/data/web/feeds", use_sudo=True, sudo_user="nginx")

atexit.register(ttrss.clean_lock)
if options["restart"]:
    atexit.register(Program.runprogram, shlex.split("systemctl restart ttrssd"), use_sudo=True)

'''
def ttrssgcfile():
    if not is_privileged():
        Program.runprogram(["touch", "/var/cache/ttrssgc"], use_sudo=True, user="root")
    else:
        pathlib.Path("/var/cache/ttrssgc").touch()
'''

if not pathlib.Path("/var/cache/ttrssgc").is_file():
    touchfile('/var/cache/ttrssgc')


def ttrssgc(then):
    diff = datediff(then)
    if diff.days > 30:
        ttrss.gc(aggressive=options["aggressive"])
        touchfile('/var/cache/ttrssgc')


if pathlib.Path("/data/web/feeds/.git").is_dir():
    ttrss.clean_lock()
    ttrss.pull()
elif pathlib.Path("/data/web/feeds").exists() and not pathlib.Path("/data/web/feeds/.git").exists():
    if options["replace"]:
        # Program.runprogram(["mv", "/data/web/feeds", "/data/web/feeds.old"], use_sudo=True, user="nginx")
        Program.runprogram(shlex.split("mv /dev/web/feeds /data/web/feeds.old"), use_sudo=True, user="nginx")
        ttrss.clone("https://tt-rss.org/gitlab/fox/tt-rss.git")
    else:
        print(Mood.sad("No git repository located in the target directory."))
        raise FileNotFoundError
elif pathlib.Path("/data/web/feeds").is_file():
    print(Mood.sad("Target location is a file, renaming and cloning repository."))
    # Program.runprogram(["mv", "/data/web/feeds", "/data/web/feeds.bad"], use_sudo=True, user="nginx")
    Program.runprogram(shlex.split("mv /data/web/feeds /data/web/feeds.bad"), use_sudo=True, user="nginx")
    ttrss.clone("https://tt-rss.org/gitlab/fox/tt-rss.git")
else:
    print(Mood.neutral("{Target location does not exist, cloning repository."))
    ttrss.clone("https://tt-rss.org/gitlab/fox/tt-rss.git")

ttrssgc(pathlib.Path("/var/cache/ttrssgc").stat().st_mtime)
